presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java

```diff
@@ -589,6 +563,171 @@ public Signature resolveFunction(QualifiedName name, List<TypeSignature> paramet
+    {
+        List<ApplicableFunction> mostSpecificFunctions = selectMostSpecificFunctions(applicableFunctions);
+        if (mostSpecificFunctions.size() > 1 && someParameterIsUnknown(parameters)) {
+            // if there is more than one function, look for functions that only cast the unknown arguments
+            List<ApplicableFunction> unknownOnlyCastFunctions = getUnknownOnlyCastFunctions(applicableFunctions, parameters);
+            if (!unknownOnlyCastFunctions.isEmpty()) {
+                mostSpecificFunctions = unknownOnlyCastFunctions;
+            }
+            if (mostSpecificFunctions.size() == 1) {
+                return mostSpecificFunctions;
+            }
+            // If the return type for all the selected function is the same, and the parameters are not declared as nullable
+            // all the functions are semantically the same. We can return just any of those.
+            if (returnTypeIsTheSame(mostSpecificFunctions) && allReturnNullOnGivenInputTypes(mostSpecificFunctions, parameters)) {
+                // make it deterministic
+                ApplicableFunction selectedFunction = Ordering.usingToString()
+                        .reverse()
+                        .sortedCopy(mostSpecificFunctions)
+                        .get(0);
+                return ImmutableList.of(selectedFunction);
+            }
+        }
+        return mostSpecificFunctions;
+    }
+
+    private List<ApplicableFunction> selectMostSpecificFunctions(List<ApplicableFunction> candidates)
+    {
+        List<ApplicableFunction> representatives = new ArrayList<>();
+
+        for (ApplicableFunction current : candidates) {
+            boolean found = false;
+            for (int i = 0; i < representatives.size(); i++) {
+                ApplicableFunction representative = representatives.get(i);
+                if (current.isMoreSpecificThan(representative)) {
+                    representatives.set(i, current);
+                }
+                if (current.isMoreSpecificThan(representative) || representative.isMoreSpecificThan(current)) {
+                    found = true;
+                    break;
+                }
+            }
+
+            if (!found) {
+                representatives.add(current);
+            }
+        }
+
+        return representatives;
+    }
+
+    private boolean someParameterIsUnknown(List<Type> parameters)
+    {
+        return parameters.stream().anyMatch(type -> type.equals(UNKNOWN));
+    }
+
+    private List<ApplicableFunction> getUnknownOnlyCastFunctions(List<ApplicableFunction> applicableFunction, List<Type> actualParameters)
+    {
+        return applicableFunction.stream()
+                .filter((function) -> onlyCastsUnknown(function, actualParameters))
+                .collect(toImmutableList());
+    }
+
+    private boolean onlyCastsUnknown(ApplicableFunction applicableFunction, List<Type> actualParameters)
```

Comment:
Just to clarify, does "only casts unknown" refer to whether only "unknown" types are being *implicitly* coerced?